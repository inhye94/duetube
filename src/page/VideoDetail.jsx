import { useQuery } from "@tanstack/react-query";
import React from "react";
import { useParams } from "react-router-dom";
import Channel from "../component/Channel";
import VideoDescription from "../component/VideoDetail/VideoDescription";
import VideoSummary from "../component/VideoDetail/VideoSummary";
import VideoViewer from "../component/VideoDetail/VideoViewer";
import { useYoutubeApi } from "../context/YoutubeApiContext";

export default function VideoDetail() {
  const { videoId } = useParams();
  const { youtube } = useYoutubeApi();

  const { isLoading, error, data } = useQuery({
    queryKey: ["detail", videoId],
    queryFn: async () => await youtube.detailAndChannel(videoId),
  });

  if (isLoading)
    return (
      <p className="loading" data-testid="detail-loading">
        Loading!!!
      </p>
    );
  if (error)
    return (
      <p className="loading" data-testid="detail-error">
        에러 났서요!
      </p>
    );

  return (
    <div className="md:flex">
      <div className="w-full md:w-3/4 md:mr-[24px]">
        {data.detail && (
          <section data-testid="detail-container">
            <header>
              <div className="relative w-full aspect-video object-contain mb-[16px]">
                <VideoViewer videoId={videoId} />
              </div>

              <div className="my-[24px]">
                <VideoSummary
                  detail={data.detail}
                  channelThumb={data.channel.snippet.thumbnails}
                />
              </div>
            </header>

            <div className="p-[16px] mb-[24px] rounded-xl bg-slate-100 dark:bg-slate-600">
              <VideoDescription detail={data.detail} />
            </div>
          </section>
        )}
      </div>

      <div className="w-full md:w-1/4">
        {data.channel && <Channel channel={data.channel} />}
      </div>
    </div>
  );
}
