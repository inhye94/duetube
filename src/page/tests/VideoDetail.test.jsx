import {
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/dom";
import { render } from "@testing-library/react";
import React from "react";
import { Route } from "react-router-dom";
import { withAllContexts, withRouter } from "../../tests/utils";
import { mockChannel, mockDetail } from "../../tests/videos";
import VideoDetail from "../VideoDetail";

// mocking
const mockVideoViewer = jest.fn();
jest.mock("../../component/VideoDetail/VideoViewer", () => {
  return ({ videoId }) => {
    mockVideoViewer(videoId);
    return <p data-testid="video-viewer" />;
  };
});

const mockVideoSummary = jest.fn();
jest.mock("../../component/VideoDetail/VideoSummary", () => {
  return ({ detail, channelThumb }) => {
    mockVideoSummary({ detail, channelThumb });
    return <div data-testid="video-summary" />;
  };
});

const mockVideoDescription = jest.fn();
jest.mock("../../component/VideoDetail/VideoDescription", () => {
  return ({ detail }) => {
    mockVideoDescription(detail);
    return <p data-testid="video-description" />;
  };
});

const mockChannelFunc = jest.fn();
jest.mock("../../component/Channel", () => {
  return ({ channel }) => {
    mockChannelFunc(channel);
    return <p data-testid="channel-detail" />;
  };
});

// test
describe("VideoDetail 컴포넌트", () => {
  // API instance를 모킹
  const fakeYoutube = { detailAndChannel: jest.fn() };
  const videoId = "12345";

  afterEach(() => {
    fakeYoutube.detailAndChannel.mockReset();
  });

  it("정상적으로 렌더링합니다", async () => {
    fakeYoutube.detailAndChannel.mockImplementation(() => ({
      detail: mockDetail,
      channel: mockChannel,
    }));
    const { asFragment } = renderVideoDetail(videoId);

    await waitForElementToBeRemoved(screen.queryByTestId("detail-loading"));

    expect(asFragment()).toMatchSnapshot();
  });

  it("정상적으로 데이터를 렌더링합니다", async () => {
    fakeYoutube.detailAndChannel.mockResolvedValue({
      detail: mockDetail,
      channel: mockChannel,
    });
    renderVideoDetail(videoId);

    await waitFor(() => screen.getByTestId("detail-container"));

    expect(fakeYoutube.detailAndChannel).toHaveBeenCalledWith(videoId);
    expect(screen.queryByTestId("detail-loading")).toBeNull();
    expect(screen.queryByTestId("detail-error")).toBeNull();
    expect(mockVideoViewer).toHaveBeenCalledWith(videoId);
    expect(mockVideoSummary).toHaveBeenCalledWith({
      detail: mockDetail,
      channelThumb: mockChannel.snippet.thumbnails,
    });
    expect(mockVideoDescription).toHaveBeenCalledWith(mockDetail);
    expect(mockChannelFunc).toHaveBeenCalledWith(mockChannel);
  });

  it("로딩 상태를 보여줍니다", () => {
    fakeYoutube.detailAndChannel.mockImplementation(() => mockDetail);
    renderVideoDetail(videoId);

    expect(screen.queryByTestId("detail-loading")).toBeInTheDocument();
  });

  it("에러 메시지를 보여줍니다", async () => {
    fakeYoutube.detailAndChannel.mockRejectedValue(new Error("에러 났서요!"));
    renderVideoDetail(videoId);

    await screen.findByTestId("detail-error");
    expect(screen.getByTestId("detail-error")).toHaveTextContent(
      "에러 났서요!"
    );
  });

  function renderVideoDetail(videoId) {
    return render(
      withAllContexts(
        withRouter(
          <Route path="/watch/:videoId" element={<VideoDetail />} />,
          `/watch/${videoId}`
        ),
        fakeYoutube
      )
    );
  }
});
