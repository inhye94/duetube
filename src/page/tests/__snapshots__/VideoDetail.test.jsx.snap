// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail 컴포넌트 정상적으로 렌더링합니다 1`] = `
<DocumentFragment>
  <div
    class="md:flex"
  >
    <div
      class="w-full md:w-3/4 md:mr-[24px]"
    >
      <section
        data-testid="detail-container"
      >
        <header>
          <div
            class="relative w-full aspect-video object-contain mb-[16px]"
          >
            <p
              data-testid="video-viewer"
            />
          </div>
          <div
            class="my-[24px]"
          >
            <div
              data-testid="video-summary"
            />
          </div>
        </header>
        <div
          class="p-[16px] mb-[24px] rounded-xl bg-slate-100 dark:bg-slate-600"
        >
          <p
            data-testid="video-description"
          />
        </div>
      </section>
    </div>
    <div
      class="w-full md:w-1/4"
    >
      <p
        data-testid="channel-detail"
      />
    </div>
  </div>
</DocumentFragment>
`;
